idb/postgres/postgres_integration_test.go

```diff
@@ -1783,3 +1793,93 @@ func TestKeytypeDoNotResetReceiver(t *testing.T) {
 	assertKeytype(t, db, test.AccountA, &keytype)
 	assertKeytype(t, db, test.AccountB, &keytype)
 }
+
+// Test that if information in `txn` and `txn_participation` tables is ahead of
+// the current round, AddBlock() still runs successfully.
+func TestAddBlockTxnTxnParticipationAhead(t *testing.T) {
+	block := test.MakeGenesisBlock()
+	db, shutdownFunc := setupIdb(t, test.MakeGenesis(), block)
+	defer shutdownFunc()
+
+	{
+		query := `INSERT INTO txn (round, intra, typeenum, asset, txn, extra)
+			VALUES (1, 0, 0, 0, 'null'::jsonb, 'null'::jsonb)`
+		_, err := db.db.Exec(context.Background(), query)
+		require.NoError(t, err)
+	}
+	{
+		query := `INSERT INTO txn_participation (addr, round, intra)
+			VALUES ($1, 1, 0)`
+		_, err := db.db.Exec(context.Background(), query, test.AccountA[:])
+		require.NoError(t, err)
+	}
+
+	txn := test.MakePaymentTxn(
+		0, 0, 0, 0, 0, 0, test.AccountA, test.AccountA, basics.Address{}, basics.Address{})
+	block, err := test.MakeBlockForTxns(block.BlockHeader, &txn)
+	require.NoError(t, err)
+	err = db.AddBlock(&block)
+	require.NoError(t, err)
+}
+
+// Test that AddBlock() writes to `txn_participation` table.
+func TestAddBlockTxnParticipationAdded(t *testing.T) {
+	block := test.MakeGenesisBlock()
+	db, shutdownFunc := setupIdb(t, test.MakeGenesis(), block)
+	defer shutdownFunc()
+
+	txn := test.MakePaymentTxn(
+		0, 0, 0, 0, 0, 0, test.AccountA, test.AccountA, basics.Address{}, basics.Address{})
+	block, err := test.MakeBlockForTxns(block.BlockHeader, &txn)
+	require.NoError(t, err)
+	err = db.AddBlock(&block)
+	require.NoError(t, err)
+
+	tf := idb.TransactionFilter{
+		Address: test.AccountA[:],
+	}
+	rowsCh, _ := db.Transactions(context.Background(), tf)
+
+	row, ok := <-rowsCh
+	require.True(t, ok)
+	require.NoError(t, row.Error)
+	require.NotNil(t, row.Txn)
+	assert.Equal(t, txn, *row.Txn)
+}
+
+// Test that if information in the `txn` table is ahead of the current round,
+// Transactions() doesn't return the rows ahead of the state.
+func TestTransactionsTxnAhead(t *testing.T) {
```

Comment:
How do these cases actually happening? Are the txn/txn_participation being added in a separate transaction? I thought it was in a nested transaction, but the test suggests otherwise.

Reply:
No, there are no nested transactions. We need to have multiple database sessions if we want to have parallelism.

Reply:
Thanks, I think this is what I was missing earlier. Your decisions here make a lot more sense to me now.